regex module 

In [ ]:
import re

sample_text = "123abc456abc789ABC"

pattern = re.compile(r'abc')   # r means raw string
matches = pattern.finditer(sample_text)
for match in matches:
    print(match)
    print(match.group())
    print(match.group(), match.start(), match.end())

<re.Match object; span=(3, 6), match='abc'>
abc
abc 3 6
<re.Match object; span=(9, 12), match='abc'>
abc
abc 9 12


to ingore case sensitive 

In [6]:
import re

sample_text = "123abc456abc789ABC"

pattern = re.compile(r'abc', re.IGNORECASE)
matches = pattern.finditer(sample_text)
for match in matches:
    print(match)
    print(match.group())
    print(match.group(), match.start(), match.end())

<re.Match object; span=(3, 6), match='abc'>
abc
abc 3 6
<re.Match object; span=(9, 12), match='abc'>
abc
abc 9 12
<re.Match object; span=(15, 18), match='ABC'>
ABC
ABC 15 18


Why use re.compile()?
When you use raw regex functions like re.search() or re.match() with pattern strings, the regex engine compiles the pattern every time the function is called.

With re.compile(), the regex pattern is compiled once into a regex object.

This compiled object can then be reused multiple times efficiently.

Especially useful when the same pattern will be used on multiple strings or repeatedly in a loop.

In [8]:
import re

sample_text = "123abc456abc789ABC"


matches = re.finditer(r'abc' ,   sample_text, re.IGNORECASE)
for match in matches:
    print(match)
    print(match.group())
    print(match.group(), match.start(), match.end())

<re.Match object; span=(3, 6), match='abc'>
abc
abc 3 6
<re.Match object; span=(9, 12), match='abc'>
abc
abc 9 12
<re.Match object; span=(15, 18), match='ABC'>
ABC
ABC 15 18


there are three main methods
1.match  # it works if the expression matches begining of the string
2.search  # where as search will check entire stinf and if any matrches found it will provide

3.find all

In [9]:
import re

sample_text = "123abc456abc789ABC"

match=re.match(r'123', sample_text)
if match:
    print("Match found:", match.group())

Match found: 123


In [10]:
import re

sample_text = "123abc456abc789ABC"
match=re.match(r'abc', sample_text)
if match:
    print("Match found:", match.group())
else:
    print("No match found")

No match found


In [11]:
import re

sample_text = "123abc456abc789ABC"
match=re.search(r'abc', sample_text)
if match:
    print("Match found:", match.group())
else:
    print("No match found")

Match found: abc


In [14]:
import re

sample_text = "123abc456abc789ABC"
match=re.findall(r'abc', sample_text, re.IGNORECASE)
if match:
    print("Match found:", match)
else:
    print("No match found")

Match found: ['abc', 'abc', 'ABC']


In [19]:
import re

sample_text = "123abc456abc789ABC"
matches=re.finditer(r'abc', sample_text, re.IGNORECASE)
for match in matches:
    print("Match found:", match)
    print(match.group(), match.start(), match.end(),match.span())

Match found: <re.Match object; span=(3, 6), match='abc'>
abc 3 6 (3, 6)
Match found: <re.Match object; span=(9, 12), match='abc'>
abc 9 12 (9, 12)
Match found: <re.Match object; span=(15, 18), match='ABC'>
ABC 15 18 (15, 18)


Metacharacters are characters with a special meaning:
All meta characters: . ^ $ * + ? { } [ ] \ | ( )
Meta characters need need to be escaped (with ) if we actually want to search for the char.

. Any character (except newline character) "he..o"
^ Starts with "^hello"
\$ Ends with "world\$"
* Zero or more occurrences "aix*"
+ One or more occurrences "aix+"
{ } Exactly the specified number of occurrences "al{2}"
[] A set of characters "[a-m]"
\ Signals a special sequence (can also be used to escape special characters) "\d"
| Either or "falls|stays"
( ) Capture and group
test_string = 'python-engineer.com'
pattern = re.compile(r'\.')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)
    <re.Match object; span=(15, 16), match='.'>
More Metacharacters / Special Sequences¶
A special sequence is a \ followed by one of the characters in the list below, and has a special meaning:

\d :Matches any decimal digit; this is equivalent to the class [0-9].
\D : Matches any non-digit character; this is equivalent to the class [^0-9].
\s : Matches any whitespace character;
\S : Matches any non-whitespace character;
\w : Matches any alphanumeric (word) character; this is equivalent to the class [a-zA-Z0-9_].
\W : Matches any non-alphanumeric character; this is equivalent to the class [^a-zA-Z0-9_].
\b Returns a match where the specified characters are at the beginning or at the end of a word r"\bain" r"ain\b"
\B Returns a match where the specified characters are present, but NOT at the beginning (or at the end) of a word r"\Bain" r"ain\B"
\A Returns a match if the specified characters are at the beginning of the string "\AThe"
\Z Returns a match if the specified characters are at the end of the string "Spain\Z"
test_string = 'hello 123_ heyho hohey'
pattern = re.compile(r'\d')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)

print()
pattern = re.compile(r'\s')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)

print()
pattern = re.compile(r'\w')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)

print()
pattern = re.compile(r'\bhey')
matches = pattern.finditer('heyho hohey') # ho-hey, ho\nhey are matches!
for match in matches:
    print(match)

print()
pattern = re.compile(r'\Ahello')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)

print()
pattern = re.compile(r'123_\Z')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)
    <re.Match object; span=(6, 7), match='1'>
    <re.Match object; span=(7, 8), match='2'>
    <re.Match object; span=(8, 9), match='3'>

    <re.Match object; span=(5, 6), match=' '>
    <re.Match object; span=(10, 11), match=' '>
    <re.Match object; span=(16, 17), match=' '>

    <re.Match object; span=(0, 1), match='h'>
    <re.Match object; span=(1, 2), match='e'>
    <re.Match object; span=(2, 3), match='l'>
    <re.Match object; span=(3, 4), match='l'>
    <re.Match object; span=(4, 5), match='o'>
    <re.Match object; span=(6, 7), match='1'>
    <re.Match object; span=(7, 8), match='2'>
    <re.Match object; span=(8, 9), match='3'>
    <re.Match object; span=(9, 10), match='_'>
    <re.Match object; span=(11, 12), match='h'>
    <re.Match object; span=(12, 13), match='e'>
    <re.Match object; span=(13, 14), match='y'>
    <re.Match object; span=(14, 15), match='h'>
    <re.Match object; span=(15, 16), match='o'>
    <re.Match object; span=(17, 18), match='h'>
    <re.Match object; span=(18, 19), match='o'>
    <re.Match object; span=(19, 20), match='h'>
    <re.Match object; span=(20, 21), match='e'>
    <re.Match object; span=(21, 22), match='y'>

    <re.Match object; span=(0, 3), match='hey'>

    <re.Match object; span=(0, 5), match='hello'>
Sets¶
A set is a set of characters inside a pair of square brackets [] with a special meaning. Append multiple conditions back-to back, e.g. [aA-Z].
A ^ (caret) inside a set negates the expression.
A - (dash) in a set specifies a range if it is in between, otherwise the dash itself.

Examples:
- [arn] Returns a match where one of the specified characters (a, r, or n) are present
- [a-n] Returns a match for any lower case character, alphabetically between a and n
- [^arn] Returns a match for any character EXCEPT a, r, and n
- [0123] Returns a match where any of the specified digits (0, 1, 2, or 3) are present
- [0-9] Returns a match for any digit between 0 and 9
- [0-5][0-9] Returns a match for any two-digit numbers from 00 and 59
- [a-zA-Z] Returns a match for any character alphabetically between a and z, lower case OR upper case

test_string = 'hello 123_'
pattern = re.compile(r'[a-z]')
matches = pattern.finditer(test_string)
for match in matches:
    print(match)
    <re.Match object; span=(0, 1), match='h'>
    <re.Match object; span=(1, 2), match='e'>
    <re.Match object; span=(2, 3), match='l'>
    <re.Match object; span=(3, 4), match='l'>
    <re.Match object; span=(4, 5), match='o'>

In [22]:
import re

sample_text = "abc123xyz456"
pattern = re.compile(r'.')
matches = pattern.finditer(sample_text)
for i in matches:
    print(i)

<re.Match object; span=(0, 1), match='a'>
<re.Match object; span=(1, 2), match='b'>
<re.Match object; span=(2, 3), match='c'>
<re.Match object; span=(3, 4), match='1'>
<re.Match object; span=(4, 5), match='2'>
<re.Match object; span=(5, 6), match='3'>
<re.Match object; span=(6, 7), match='x'>
<re.Match object; span=(7, 8), match='y'>
<re.Match object; span=(8, 9), match='z'>
<re.Match object; span=(9, 10), match='4'>
<re.Match object; span=(10, 11), match='5'>
<re.Match object; span=(11, 12), match='6'>


In [25]:
import re

sample_text = "abc123xyz456"
pattern=r'.'
matches = re.finditer(pattern,sample_text)
for i in matches:
    print(i.group())

a
b
c
1
2
3
x
y
z
4
5
6


In [27]:
import re

sample_text = "abc123xyz456."
pattern=r'\.'  # Escape the dot to match a literal dot
matches = re.finditer(pattern,sample_text)
for i in matches:
    print(i)

<re.Match object; span=(12, 13), match='.'>


In [ ]:
import re

sample_text = "abc123xyz456."
pattern=r'^123'  
matches = re.finditer(pattern,sample_text)
for i in matches:
    print(i) # No output since '123' is not at the start

In [29]:
import re

sample_text = "abc123xyz456."
pattern=r'^abc'  
matches = re.finditer(pattern,sample_text)
for i in matches:
    print(i.group()) #output since 'abc' is at the start

abc


In [30]:
import re

sample_text = "abc123xyz456."
pattern=r'abc$'  # Match 'abc' at the end of the string
matches = re.finditer(pattern,sample_text)
for i in matches:
    print(i.group()) # No output since 'abc' is not at the end

In [31]:
import re

sample_text = "abc123xyz456."
pattern=r'456$' # Match '456' at the end of the string
matches = re.finditer(pattern,sample_text)
for i in matches:
    print(i.group()) #output since '456' is at the end